RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc

```diff
@@ -0,0 +1,199 @@
+      PFRecHitDeviceCollection::View::element pfrh, const ECAL::CaloRecHitSoATypeDevice::ConstView::const_element rh) {
+    pfrh.detId() = rh.detId();
+    pfrh.energy() = rh.energy();
+    pfrh.time() = rh.time();
+    pfrh.depth() = 1;
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == EcalBarrel)
+      pfrh.layer() = PFLayer::ECAL_BARREL;
+    else if (subdet == EcalEndcap)
+      pfrh.layer() = PFLayer::ECAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  template <typename CAL>
+  struct PFRecHitProducerKernelImpl2 {
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const typename CAL::TopologyTypeDevice::ConstView topology,
+                                  PFRecHitDeviceCollection::View pfRecHits,
+                                  const uint32_t* __restrict__ denseId2pfRecHit,
+                                  uint32_t* __restrict__ num_pfRecHits) const {
+      // First thread updates size field pfRecHits SoA
+      if (const int32_t thread = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]; thread == 0)
+        pfRecHits.size() = *num_pfRecHits;
+
+      // Assign position information and associate neighbours
+      for (int32_t i : cms::alpakatools::elements_with_stride(acc, *num_pfRecHits)) {
+        const uint32_t denseId = CAL::detId2denseId(pfRecHits.detId(i));
+
+        pfRecHits.x(i) = topology.positionX(denseId);
+        pfRecHits.y(i) = topology.positionY(denseId);
+        pfRecHits.z(i) = topology.positionZ(denseId);
+
+        for (uint32_t n = 0; n < 8; n++) {
+          pfRecHits.neighbours(i)(n) = -1;
+          const uint32_t denseId_neighbour = topology.neighbours(denseId)(n);
+          if (denseId_neighbour != 0xffffffff) {
+            const uint32_t pfRecHit_neighbour = denseId2pfRecHit[denseId_neighbour];
+            if (pfRecHit_neighbour != 0xffffffff)
+              pfRecHits.neighbours(i)(n) = (int32_t)pfRecHit_neighbour;
+          }
+        }
+      }
+    }
+  };
+
+  template <typename CAL>
+  PFRecHitProducerKernel<CAL>::PFRecHitProducerKernel(Queue& queue)
+      : denseId2pfRecHit_(cms::alpakatools::make_device_buffer<uint32_t[]>(queue, CAL::SIZE)),
+        num_pfRecHits_(cms::alpakatools::make_device_buffer<uint32_t>(queue)),
+        work_div_(cms::alpakatools::make_workdiv<Acc1D>(1, 1)) {}
+
+  template <typename CAL>
+  void PFRecHitProducerKernel<CAL>::prepareEvent(Queue& queue, const uint32_t num_recHits) {
+    alpaka::memset(queue, denseId2pfRecHit_, 0xff);  // Reset denseId -> pfRecHit index map
+    alpaka::memset(queue, num_pfRecHits_, 0x00);     // Reset global pfRecHit counter
+
+    // Use only one block on the synchronous CPU backend, because there is no
+    // performance gain in using multiple blocks, but there is a significant
+    // penalty due to the more complex synchronisation.
```

Comment:
Could you elaborate this comment? CPU serial backend should just loop serially over the blocks, and thus not have any special synchronization. Or does the comment anticipate a CPU parallel backend (e.g. TBB)?

Reply:
The first kernel requires global synchronisation, which is slow when using multiple blocks. When using the serial CPU backend, there is no performance advantage in using multiple blocks, but a significant penalty due to the synchronisation. Therefore, I use a single block for the CPU backend and corresponding synchronisation (this also addresses you `::Blocks` vs `::Grids` comment above). But you are right, the CPU backend does not require any synchronisation, so I can skip the atomic operation in this case.

Reply:
I'm actually still confused of my `::Blocks` vs `::Grids` question :) IIUC you want the `j` on line 27 to have unique values for every thread within a grid. IIUC this would correspond the `::Blocks` setting. AFAICT for CPU Serial the `::Blocks` synchronization level on atomics has zero cost (whereas `::Grids` has a cost). Or do you really need to synchronize across grids?

Reply:
Yes, `j` is the index of the newly constructed PFRecHit. So it has to be unique and consecutive across all threads. That is why the kernel requires global synchronisation, which is achieved with the `::Grids` parameter in the atomic operation. I found that the `::Grids` synchronisation has a significant cost in the serial CPU backend (as you say).
My solution was to use a single block for the CPU backend and switch to `::Blocks` synchronisation. The synchronisation is not necessary, logically it should be there (and you say it has zero cost).
Alternatively, one could skip the atomic operation for the CPU backend altogether (since it is serial anyway) and use multiple blocks also for the CPU. This would however not work with a potential parallel CPU backend.

I suggest to do the following: Keep the layouts and synchronisation as they are, but replace `cms::alpakatools::requires_single_thread_per_block_v<Acc1D>` by `std::is_same_v<Device, alpaka::DevCpu>` in lines 27 and 162. For the serial CPU backend, it makes no difference anyway, but this should also work reasonably with a potential parallel CPU backend.

Reply:
> Yes, `j` is the index of the newly constructed PFRecHit. So it has to be unique and consecutive across all threads. That is why the kernel requires global synchronisation, which is achieved with the `::Grids` parameter in the atomic operation

Shouldn't `::Blocks` be sufficient to provide synchronization across threads of a block?

Reply:
Yes, but the GPU backend uses multiple blocks. Synchronisation within each block is not sufficient. That is why we need the more complex `::Grids` synchronisation there.
For the CPU backend, we use a single block, so `::Blocks` is appropriate.

Reply:
I fully agree this piece needs synchronization across blocks. But my impression of the Alpaka atomic operation synchronization hierarchy (from quick poking of alpaka code few weeks ago) is
* `::Theads` provides synchronization across threads of a block
* `::Blocks` provides synchronization across (threads of) blocks of a grid (i.e. kernel)
* `::Grids` provides synchronization across (threads of blocks of) grids (i.e. kernels)

If this is the case, then I'd think `::Blocks` would be sufficient for your case. Do you have any pointer to documentation or experience that the behavior of `::Blocks` is something different?

Reply:
I am not sure where I got the `::Blocks` and `::Grids` from and I also cannot find any documentation right now. If what you say is correct, `::Blocks` should be changed to `::Threads` for the CPU and `::Grids` to `::Blocks` for the GPU.
I tried changing `::Grids` to `::Blocks` and ran lots of validation of the GPU backend. It validated fine, so you are probably correct. (Then again, I also tried using `::Threads` (which should not work) and it still validated ok, so I am not completely sure)

Reply:
(assuming my understanding of the atomic operation hierarchy levels is correct) I'd suggest to just use `::Blocks` for all backends, i.e. remove the special treatment of the CPU backend.

Reply:
@fllor What do you think about @makortel 's suggestions?